### PR TITLE
Add testcases for SSL_key_update problems

### PR DESCRIPTION
**Pull request by yangyangtiantianlonglong**
Add testcases for SSL_key_update() corner case calls

Test that SSL_key_update() is not allowed if there are writes pending.
Test that there is no reset of the packet pointer in ssl3_setup_read_buffer().

Fix https://github.com/openssl/openssl/pull/16077
Discuss https://github.com/openssl/openssl/pull/16021

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by t8m** [comment]
Could you please reword the commit message a little bit?
Something like:
```
Add testcases for SSL_key_update() corner case calls

Test that SSL_key_update() is not allowed if there are writes pending.
Test that there is no reset of the packet pointer in ssl3_setup_read_buffer().
```

File: doc/man3/SSL_get_error.pod [hunk 1]
```diff
@@ -85,7 +85,9 @@ protocol level.
 It is safe to call SSL_read() or SSL_read_ex() when more data is available
 even when the call that set this error was an SSL_write() or SSL_write_ex().
 However, if the call was an SSL_write() or SSL_write_ex(), it should be called
-again to continue sending the application data.
+again to continue sending the application data.if you get B<SSL_ERROR_WANT_WRITE>
```
Comment (t8m) [inline doc/man3/SSL_get_error.pod:88]: `...data. If you...`

File: doc/man3/SSL_get_error.pod [hunk 2]
```diff
@@ -85,7 +85,9 @@ protocol level.
...
+from SSL_write() or SSL_write_ex() then you should not do any other B<IO> operation
```
Comment (t8m) [inline doc/man3/SSL_get_error.pod:89]: `then you should not do any other operation that could trigger IO other than to repeat the previous SSL_write() call.`

**Review by t8m — CHANGES_REQUESTED** [review]

File: test/sslapitest.c [hunk 3]
```diff
@@ -6192,6 +6192,262 @@ static int test_key_update_in_write(int tst)
...
+
+    SSL_set_bio(local, lbio, lbio);
+    SSL_set_bio(peer, pbio, pbio);
+
+    /*
+     * we first write keyupdate msg then appdata in local
+     * write data in local will fail with SSL_ERROR_WANT_WRITE
+     */
+    if (!TEST_true(SSL_key_update(local, SSL_KEY_UPDATE_REQUESTED))
+            || !TEST_int_eq(SSL_write(local, lwbuf, sizeof(lwbuf)), -1)
```
Comment (mattcaswell) [inline test/sslapitest.c:6238]: A comment would be useful here. It is not immediately obvious why this causes SSL_ERROR_WANT_WRITE, i.e. because lwbuf app data msg size + key update msg size > 512 (the size of the bio pair buffer)

File: test/sslapitest.c [hunk 4]
```diff
@@ -6192,6 +6192,262 @@ static int test_key_update_in_write(int tst)
...
+    if (!TEST_false(SSL_key_update(local, SSL_KEY_UPDATE_REQUESTED))
+        || !TEST_int_eq(ERR_GET_REASON(ERR_peek_error()), SSL_R_BAD_WRITE_RETRY))
+    goto end;
+
+    ERR_clear_error();
+    /* write data in local previously that we will complete */
+    if (!TEST_int_eq(SSL_write(local, mess, strlen(mess)), strlen(mess)))
+        goto end;
+
+    /* SSL_key_update will sucess, write data previously already complete */
```
Comment (mattcaswell) [inline test/sslapitest.c:6338]: SSL_key_update will succeed because there is no pending write data

File: test/sslapitest.c [hunk 5]
```diff
@@ -6192,6 +6192,262 @@ static int test_key_update_in_write(int tst)
...
+
+    /*
+     * we write some appdata in local
+     * read data in peer - we will read the keyupdate msg
+     */
+    if (!TEST_int_eq(SSL_write(local, mess, strlen(mess)), strlen(mess))
+        || !TEST_int_eq(SSL_read(peer, buf, sizeof(buf)), strlen(mess)))
+        goto end;
+
+    /* Write in peer more data to ensure we send the keyupdate message back */
```
Comment (mattcaswell) [inline test/sslapitest.c:6351]: Write more peer data...

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: doc/man3/SSL_get_error.pod (hunk 2 above)
Comment (yangyangtiantianlonglong) [reply doc/man3/SSL_get_error.pod]: ok, Modified

File: doc/man3/SSL_get_error.pod (hunk 1 above)
Comment (yangyangtiantianlonglong) [reply doc/man3/SSL_get_error.pod]: ok, Modified

File: test/sslapitest.c (hunk 3 above)
Comment (yangyangtiantianlonglong) [reply test/sslapitest.c:6240]: ok, Modified

File: test/sslapitest.c (hunk 4 above)
Comment (yangyangtiantianlonglong) [reply test/sslapitest.c]: ok, Modified

File: test/sslapitest.c (hunk 5 above)
Comment (yangyangtiantianlonglong) [reply test/sslapitest.c]: ok, Modified

**Comment by yangyangtiantianlonglong** [comment]
I have changed the early review opinions, I can't understand what meaning is outdated?

**Comment by kaduk** [comment]
> I can't understand what meaning is outdated?

github says that the review comment is "outdated" when the comment was made on a version of the code (or, in this case, documentation) that is not the current version of the pull request.  In that sense the comment is "out of date", and thus "outdated".

**Review by t8m — APPROVED** [review]

**Review by paulidale — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by paulidale** [comment]
Merged, thanks for this.
